processing/src/main/java/org/apache/druid/segment/serde/CompressedComplexColumnSerializer.java

```diff
@@ -0,0 +1,164 @@
+import java.nio.ByteOrder;
+import java.nio.channels.WritableByteChannel;
+
+public class CompressedComplexColumnSerializer<T> implements GenericColumnSerializer<T>
+{
+  public static final byte IS_COMPRESSED = Byte.MAX_VALUE;
+  public static final byte V0 = 0x00;
+  public static final String FILE_NAME = "__complexColumn";
+
+  public static GenericColumnSerializer create(
+      SegmentWriteOutMedium segmentWriteOutMedium,
+      String column,
+      IndexSpec indexSpec,
+      ObjectStrategy objectStrategy
+  )
+  {
+    return new CompressedComplexColumnSerializer(column, segmentWriteOutMedium, indexSpec, objectStrategy);
+  }
+
+  public CompressedComplexColumnSerializer(
+      String name,
+      SegmentWriteOutMedium segmentWriteOutMedium,
+      IndexSpec indexSpec,
+      ObjectStrategy<T> strategy
+  )
+  {
+    this.name = name;
+    this.segmentWriteOutMedium = segmentWriteOutMedium;
+    this.indexSpec = indexSpec;
+    this.strategy = strategy;
+  }
+
+  private final String name;
+  private final SegmentWriteOutMedium segmentWriteOutMedium;
+  private final IndexSpec indexSpec;
+  private final ObjectStrategy<T> strategy;
+  private CompressedVariableSizedBlobColumnSerializer writer;
+  private ByteBufferWriter<ImmutableBitmap> nullBitmapWriter;
+  private MutableBitmap nullRowsBitmap;
+
+  private int rowCount = 0;
+  private boolean closedForWrite = false;
+  private byte[] metadataBytes;
+
+  @Override
+  public void open() throws IOException
+  {
+    writer = new CompressedVariableSizedBlobColumnSerializer(
+        NestedCommonFormatColumnSerializer.getInternalFileName(name, FILE_NAME),
+        segmentWriteOutMedium,
+        indexSpec.getComplexMetricCompression()
+    );
+    writer.open();
+
+    nullBitmapWriter = new ByteBufferWriter<>(
+        segmentWriteOutMedium,
+        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
+    );
+    nullBitmapWriter.open();
+
+    nullRowsBitmap = indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap();
+  }
+
+  @Override
+  public void serialize(ColumnValueSelector<? extends T> selector) throws IOException
+  {
+    final T data = selector.getObject();
+    if (data == null) {
+      nullRowsBitmap.add(rowCount);
+    }
+    rowCount++;
+    final byte[] bytes = strategy.toBytes(data);
+    writer.addValue(bytes);
+  }
+
+  @Override
+  public long getSerializedSize() throws IOException
+  {
+    closeForWrite();
+    // COMPRESSED_BYTE + V0 + metadata
+    return 1 + 1 + metadataBytes.length;
+  }
+
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    Preconditions.checkState(closedForWrite, "Not closed yet!");
+
+    channel.write(ByteBuffer.wrap(new byte[]{IS_COMPRESSED}));
+    channel.write(ByteBuffer.wrap(new byte[]{V0}));
+    channel.write(ByteBuffer.wrap(metadataBytes));
+
+    NestedCommonFormatColumnSerializer.writeInternal(smoosher, writer, name, FILE_NAME);
```

Comment:
yea, fair.

I think I would like `NestedCommonFormatColumnSerializer` itself to be a more neutral thing since i think the way it splits up parts into separate smoosh files is advantageous if we implement stuff like partial segment download, though I'd like to make some adjustments to smoosh metadata format to eliminate the chance for name collisions.